src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+ *      will depend on the leftmost column, this way most of the time no remote memory would need to be accessed
+ *
+ * The sorting, which is the most expensive part of this operator, could probably be parallelized.
+ */
+
+namespace {
+
+// See doc above
+using ReferenceMatrix = std::vector<opossum::PosList>;
+
+using VirtualPosList = std::vector<size_t>;
+
+/**
+ * Comparator for performing the std::sort() of a virtual pos list
+ */
+struct VirtualPosListCmpContext {
+  ReferenceMatrix& reference_matrix;
+  bool operator()(size_t lhs, size_t rhs) const {
+    for (const auto& reference_matrix_column : reference_matrix) {
+      const auto left_row_id = reference_matrix_column[lhs];
+      const auto right_row_id = reference_matrix_column[rhs];
+
+      if (left_row_id < right_row_id) return true;
+      if (right_row_id < left_row_id) return false;
+    }
+    return false;
+  }
+};
+
+}  // namespace
+
+namespace opossum {
+
+SetUnion::SetUnion(const std::shared_ptr<const AbstractOperator>& left,
+                   const std::shared_ptr<const AbstractOperator>& right)
+    : AbstractReadOnlyOperator(left, right) {}
+
+uint8_t SetUnion::num_in_tables() const { return 2; }
+
+uint8_t SetUnion::num_out_tables() const { return 1; }
+
+std::shared_ptr<AbstractOperator> SetUnion::recreate(const std::vector<AllParameterVariant>& args) const {
+  return std::make_shared<SetUnion>(input_left()->recreate(args), input_right()->recreate(args));
+}
+
+const std::string SetUnion::name() const { return "SetUnion"; }
+
+const std::string SetUnion::description() const { return "SetUnion"; }
+
+std::shared_ptr<const Table> SetUnion::_on_execute() {
+  const auto early_result = _analyze_input();
+  if (early_result) {
+    return early_result;
+  }
+
+  /**
+   * For each input, create a ReferenceMatrix
+   */
+  const auto build_reference_matrix = [&](const auto& input_table, auto& reference_matrix) {
+    reference_matrix.resize(_column_segment_begins.size());
+    for (auto& pos_list : reference_matrix) {
+      pos_list.reserve(input_table->row_count());
+    }
+
+    for (auto chunk_id = ChunkID{0}; chunk_id < input_table->chunk_count(); ++chunk_id) {
+      const auto& chunk = input_table->get_chunk(ChunkID{chunk_id});
+
+      for (size_t segment_id = 0; segment_id < _column_segment_begins.size(); ++segment_id) {
+        const auto column_id = _column_segment_begins[segment_id];
+        const auto column = chunk.get_column(column_id);
+        const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column);
+
+        auto& out_pos_list = reference_matrix[segment_id];
+        auto in_pos_list = ref_column->pos_list();
+        std::copy(in_pos_list->begin(), in_pos_list->end(), std::back_inserter(out_pos_list));
+      }
+    }
+  };
+
+  ReferenceMatrix reference_matrix_left;
+  ReferenceMatrix reference_matrix_right;
+
+  build_reference_matrix(_input_table_left(), reference_matrix_left);
```

Comment:
Why do you pass `reference_matrix_left` by reference if it is apparently the return type of that function?

Reply:
done